src/main/java/com/snowflake/kafka/connector/internal/streaming/DirectTopicPartitionChannel.java

```diff
@@ -1,82 +1,989 @@
+    Schema keySchema = isKey ? new SnowflakeJsonSchema() : record.keySchema();
+    Object keyContent = isKey ? newSFContent : record.key();
+    Schema valueSchema = isKey ? record.valueSchema() : new SnowflakeJsonSchema();
+    Object valueContent = isKey ? record.value() : newSFContent;
+    return new SinkRecord(
+            record.topic(),
+            record.kafkaPartition(),
+            keySchema,
+            keyContent,
+            valueSchema,
+            valueContent,
+            record.kafkaOffset(),
+            record.timestamp(),
+            record.timestampType(),
+            record.headers());
   }
 
+  // --------------- BUFFER FLUSHING LOGIC --------------- //
+
   @Override
-  public long getProcessedOffset() {
-    return 0;
+  public void insertBufferedRecordsIfFlushTimeThresholdReached() {
+    // It is just a leftover after buffered channel, not needed here as a buffer was removed in the current class.
+    // todo remove this method in the future
   }
 
+  public void transformAndSend(SinkRecord kafkaSinkRecord) {
+    try {
+      Map<String, Object> transformedRecord = transformDataBeforeSending(kafkaSinkRecord);
+      if (!transformedRecord.isEmpty()) {
+        InsertValidationResponse response = insertRowsWithFallback(transformedRecord, kafkaSinkRecord.kafkaOffset());
+        this.processedOffset.set(kafkaSinkRecord.kafkaOffset());
+
+        LOGGER.info(
+                "Successfully called insertRows for channel:{}, insertResponseHasErrors:{},",
+                this.getChannelNameFormatV1(),
+                response.hasErrors());
+        if (response.hasErrors()) {
+          handleInsertRowsFailure(response.getInsertErrors(), kafkaSinkRecord);
+        }
+      }
+
+    } catch (TopicPartitionChannelInsertionException ex) {
+      // Suppressing the exception because other channels might still continue to ingest
+      LOGGER.warn(
+              String.format(
+                      "[INSERT_BUFFERED_RECORDS] Failure inserting rows for channel:%s", this.getChannelNameFormatV1()),
+              ex);
+    }
+  }
+
+  /**
+   * Uses {@link Fallback} API to reopen the channel if insertRows throws {@link SFException}.
+   *
+   * <p>We have deliberately not performed retries on insertRows because it might slow down overall
+   * ingestion and introduce lags in committing offsets to Kafka.
+   *
+   * <p>Note that insertRows API does perform channel validation which might throw SFException if
+   * channel is invalidated.
+   *
+   * <p>It can also send errors {@link
+   * net.snowflake.ingest.streaming.InsertValidationResponse.InsertError} in form of response inside
+   * {@link InsertValidationResponse}
+   *
+   * @return InsertValidationResponse a response that wraps around InsertValidationResponse
+   */
+  private InsertValidationResponse insertRowsWithFallback(Map<String, Object> transformedRecord, long offset) {
+    Fallback<Object> reopenChannelFallbackExecutorForInsertRows =
```

Comment:
i know this code was here initially, but it does obscure the view a lot. what do you think - could we perhaps extract the failsafe handling outside, to some other class, make it sort of an aspect applied on top of the channel, so from channel perspective-  we simply insert rows?